File: backend/app/routers/finance.py
from fastapi import APIRouter, Depends, File, HTTPException, Query, UploadFile
from fastapi.responses import PlainTextResponse

from app.middleware.auth_middleware import CurrentUser, get_current_user
from app.schemas.finance import FinanceUploadConfirmRequest
from app.services.finance_service import (
    confirm_finance_upload,
    get_finance_billable,
    get_finance_template,
    get_upload_history,
    validate_finance_csv,
)

router = APIRouter(prefix="/finance", tags=["Finance"])


@router.get("/template", response_class=PlainTextResponse)
async def download_template(user: CurrentUser = Depends(get_current_user)):
    """Download the CSV template for finance billable uploads."""
    return PlainTextResponse(
        content=get_finance_template(),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=finance_billable_template.csv"},
    )


@router.post("/billable/upload")
async def upload_finance_csv(
    file: UploadFile = File(...),
    period: str = Query(..., description="Period in YYYY-MM format"),
    user: CurrentUser = Depends(get_current_user),
):
    """Upload and validate a finance billable CSV file."""
    if not file.filename or not file.filename.endswith(".csv"):
        raise HTTPException(status_code=400, detail="File must be a CSV")

    content = await file.read()
    result = await validate_finance_csv(content, period, user.branch_location_id)
    return result


@router.post("/billable/confirm")
async def confirm_upload(
    request: FinanceUploadConfirmRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """Confirm a previously validated finance billable upload."""
    result = await confirm_finance_upload(
        request.upload_token,
        user.user_id,
        user.branch_location_id,
    )
    if not result:
        raise HTTPException(status_code=400, detail="Invalid or expired upload token")
    return result


@router.get("/billable")
async def list_finance_billable(
    period: str = Query(..., description="Period in YYYY-MM format"),
    version: int | None = Query(None, description="Version number (defaults to latest)"),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    user: CurrentUser = Depends(get_current_user),
):
    """List finance billable entries for a period, with optional version filter."""
    return await get_finance_billable(
        period=period,
        branch_location_id=user.branch_location_id,
        version=version,
        page=page,
        page_size=page_size,
    )


@router.get("/uploads")
async def list_upload_history(
    user: CurrentUser = Depends(get_current_user),
):
    """List all finance upload history for the branch."""
    return await get_upload_history(user.branch_location_id)

File: frontend/src/api/finance.ts
import client from "./client"
import type {
  FinanceUploadValidationResponse,
  FinanceBillableListResponse,
  FinanceUploadHistoryEntry,
} from "@/types/finance"

export async function downloadFinanceTemplate(): Promise<Blob> {
  const res = await client.get("/finance/template", { responseType: "blob" })
  return res.data
}

export async function uploadFinanceCsv(
  file: File,
  period: string
): Promise<FinanceUploadValidationResponse> {
  const form = new FormData()
  form.append("file", file)
  const res = await client.post<FinanceUploadValidationResponse>(
    "/finance/billable/upload",
    form,
    {
      params: { period },
      headers: { "Content-Type": "multipart/form-data" },
    }
  )
  return res.data
}

export async function confirmFinanceUpload(
  uploadToken: string
): Promise<{ imported_count: number; version: number; message: string }> {
  const res = await client.post<{
    imported_count: number
    version: number
    message: string
  }>("/finance/billable/confirm", { upload_token: uploadToken })
  return res.data
}

export async function getFinanceBillable(params: {
  period: string
  data_source?: "hrms" | "excel"
  version?: number
  page?: number
  page_size?: number
}): Promise<FinanceBillableListResponse> {
  const res = await client.get<FinanceBillableListResponse>("/finance/billable", {
    params,
  })
  return res.data
}

export async function getFinanceUploadHistory(
  dataSource: "hrms" | "excel" = "hrms"
): Promise<FinanceUploadHistoryEntry[]> {
  const res = await client.get<FinanceUploadHistoryEntry[]>("/finance/uploads", {
    params: { data_source: dataSource },
  })
  return res.data
}
